### PR TITLE
feat: Date picking with timezones

### LINKED ISSUE #6392: Verify datepicker (budget) logic
Get to know expected behavior for time zones and fix or make sure that it is correct.

### PR DESCRIPTION
**Pull request by Zeyber**
Closes: #6392

### CODE REVIEW

**Review by Pucek9 — CHANGES_REQUESTED** [review]
While edit budget, I don't see pre-selected timezone properly (it's automatically set to +1 even it has been updated with +4)

**Comment by Zeyber** [comment]
> While edit budget, I don't see pre-selected timezone properly (it's automatically set to +1 even it has been updated with +4)

This is something we might have to live with for now since we are not able to determine any preselected timezone other than local as we cannot store any timezone data on the backend. Default selection is always local timezone (+1 in our case).

File: projects/storefrontlib/src/cms-components/organization/budgets/budget-form/budget-form.component.spec.ts [hunk 1]
```diff
@@ -202,4 +225,51 @@ describe('BudgetFormComponent', () => {
...
+      spyOn(component, 'getLocalTimezoneOffset').and.callThrough();
+      expect(component.getLocalTimezoneOffset()).toEqual('-02:00');
+    });
+
+    it('should return future offset string', () => {
+      // Use 3 hours in past timezone
+      Date.prototype.getTimezoneOffset = () => {
+        return -180;
+      };
+      spyOn(component, 'getLocalTimezoneOffset').and.callThrough();
```
Comment (Platonn) [inline projects/storefrontlib/src/cms-components/organization/budgets/budget-form/budget-form.component.spec.ts]: Do we really need this spy? (same for similar tests)

File: projects/storefrontlib/src/cms-components/organization/budgets/budget-form/budget-form.component.spec.ts [hunk 2]
```diff
@@ -202,4 +225,51 @@ describe('BudgetFormComponent', () => {
       expect(component.clickBack.emit).toHaveBeenCalledWith();
     });
   });
+
+  describe('getLocalTimezoneOffset()', () => {
+    it('should return utc-0 offset string', () => {
+      // Use UTC-0 timezone
+      Date.prototype.getTimezoneOffset = () => {
```
Comment (Platonn) [inline projects/storefrontlib/src/cms-components/organization/budgets/budget-form/budget-form.component.spec.ts]: Modifying the prototype of a global javascript object may potentially affect other tests (*also in different files*).

It would be good to clean up (bring the original prototype in place after the test executes). There are many possible solutions. I'll mention 2 (but I like the second more, as the logic is encapsulated in one place):
1. remember the original Date.prototype.getTimezoneOffset in some variable in `beforeEach` and assign it back to the Date.prototype.getTimezoneOffset in `afterEach`
2. create an util function that does similar thing, but inside itself (no need for before each and after each), i.e.
```typescript
function fakeDateTimezoneOffset(offset: number, callback: Function) {
  const original = Date.prototype.getTimezoneOffset;
  Date.prototype.getTimezoneOffset = () => offset;
  callback();
  Date.prototype.getTimezoneOffset = original;
}
```
```typescript
it('should return utc-0 offset string, fakeDateTimezoneOffset(0, () => {
   expect(component.getLocalTimezoneOffset()).toEqual('+00:00');
}))
```

What do you think about it?

File: projects/storefrontlib/src/shared/components/select-timezone/timezones.ts [hunk 3]
```diff
@@ -0,0 +1,1349 @@
+export const timezones = [
```
Comment (Platonn) [inline projects/storefrontlib/src/shared/components/select-timezone/timezones.ts]: This is a lot of additional code that will enlarge the final JS app shipped to each end user, so it slows down the user experience. Adding it should be justified. I'm interested in alternative solutions we tired.

Could you explain why we need all this data? Ideally we could get rid of them. Or at least truncate unused properties.

It looks like copied from somewhere so I'm sure you've checked the license and we can use it for free (i.e. MIT license). Usually the license requires at least the notice about the original author/owner of the copyrights. But I can't see any notice here.

File: projects/storefrontlib/src/cms-components/organization/budgets/budget-form/budget-form.component.ts [hunk 4]
```diff
@@ -56,7 +61,72 @@ export class BudgetFormComponent extends AbstractFormComponent
...
+
+      this.form.patchValue({
+        ...this.budgetData,
+        startDate: startDate,
+        endDate: endDate,
+      });
     }
   }
+
+  getLocalTimezoneOffset(invert?: boolean): string {
```
Comment (Platonn) [inline projects/storefrontlib/src/cms-components/organization/budgets/budget-form/budget-form.component.ts]: It's a complex cutom converting logic in this file. *What a problem does it solve?*

---

By the way, suggestions:
1. Ideally I'd like to avoid custom calculations in our library whenever possible, look for alternatives, maybe we can solve it different way or reuse some tool.
2. If we really need to perform them, it would be good to encapsulate it in a separate service, so it's easier testable and reusable by other components (and customers).
3. Then the methods should be well documented with examples of usage and example results. It's to help teammates and customers to undestand the API to (re)use.
4. If the problem is with converting UI model <--> Backend model, the most appropriate place for it would be converters ( OCC normalizers and serializers)

File: projects/storefrontlib/src/shared/components/select-timezone/timezones.ts (hunk 3 above)
Comment (Zeyber) [reply projects/storefrontlib/src/shared/components/select-timezone/timezones.ts]: It was there to allow for timezone selection but we are no longer using it for this implementation. I thought to leave it there for use in the future but you are right. It does not have any purpose anymore and should be removed.

File: projects/storefrontlib/src/shared/components/index.ts [hunk 5]
```diff
@@ -12,4 +13,5 @@ export * from './modal/index';
 export * from './spinner/index';
 export * from './split-view/index';
 export * from './star-rating/index';
+export * from './form/form-errors/index';
```
Comment (Platonn) [inline projects/storefrontlib/src/shared/components/index.ts]: Doesn't it raise errors? I can see the same in line 6.

File: projects/storefrontlib/src/shared/components/date-time-picker/date-time-picker-formatter.service.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
+import { Injectable } from '@angular/core';
+
+/**
+ * Service responsible for converting date-link strings to/from formats compatible with the `<input type="datetime-local">`
```
Comment (Platonn) [inline projects/storefrontlib/src/shared/components/date-time-picker/date-time-picker-formatter.service.ts]:
```suggestion
 * Service responsible for converting date-like strings to/from formats compatible with the `<input type="datetime-local">`
```

File: projects/storefrontlib/src/shared/components/date-time-picker/date-time-picker-formatter.service.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  /**
+   * Format datetime-local string into a format compatable with OCC.
+   * @param value: datetime-local string to convert
+   *
+   * @example
+   * With UTC-0 locale offset, `toDateTimeOccFormat('2010-01-01T00:00')` returns `'2010-01-01T00:00:00+00:00'`.
+   */
+  protected toDateTimeOccFormat(value: string) {
```
Comment (Platonn) [inline projects/storefrontlib/src/shared/components/date-time-picker/date-time-picker-formatter.service.ts]: Customers can use different backends than OCC (which is the default one). The layer that decides about using occ is _Occ Adapters_ in the core library. Ideally the UI-related code, including components and the whole storefront lib do not know which backend is used, so I'd avoid mentioning the phrase `OCC` here. 

I have problem to find apropriate name for this method, becuse `toISOFormat` would suggest it's `xxxZ` (like in native JS Date() object), but we mean the ISO format `+xxxx` here.

Then... what do you think about simply moving this 1-line implementation directly to the `toModel` method?

File: projects/storefrontlib/src/shared/components/date-time-picker/date-time-picker-formatter.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+          expect(service.toNative('2034-07-12T23:59-03:00')).toEqual(
+            '2034-07-13T02:59'
+          );
+        });
+      });
+    });
+  });
+
+  describe('toModel', () => {
+    it('should return undefined if value is empty', () => {
```
Comment (Platonn) [inline projects/storefrontlib/src/shared/components/date-time-picker/date-time-picker-formatter.service.spec.ts]:
```suggestion
    it('should return null if value is empty', () => {
```

File: projects/storefrontlib/src/shared/components/date-picker/date-picker-formatter.service.ts [hunk 9]
```diff
@@ -16,9 +16,9 @@ export class DatePickerFormatterService {
         .toISOString()
         .replace('.', '+')
         .replace('Z', '0');
+
       if (endOfDay) {
-        // TODO Here should be 23:59:59, but from backend I'm getting (CEST) '00:59:59', IMO backend should use UTC-0
-        date = date.replace('00:00:00', '00:59:59');
+        date = date.replace('00:00:00', '23:59:59');
```
Comment (Platonn) [inline projects/storefrontlib/src/shared/components/date-picker/date-picker-formatter.service.ts:21]: What's the reason of this change? How it is related to the current task?

File: projects/storefrontlib/src/shared/components/date-time-picker/date-time-picker-formatter.service.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      .toString()
+      .padStart(2, '0');
+    const minutes = (offset % 60).toString().padStart(2, '0');
+    return offset >= 0
+      ? !invert
+        ? `+${hours}:${minutes}`
+        : `-${hours}:${minutes}`
+      : !invert
+      ? `-${hours}:${minutes}`
+      : `+${hours}:${minutes}`;
```
Comment (Platonn) [inline projects/storefrontlib/src/shared/components/date-time-picker/date-time-picker-formatter.service.ts]: Bagatelle: As a small code optimization we could consider:
- calculating only a `sign` in the logic and leaving the repated pattern `hours:minutes` apart from the logic
- For readability it would be good to avoid negation in the condition of ternary operator (`!invert` --> `invert`)

For instance:
```typescript
const sign = offset >= 0
      ? invert
        ? `-`
        : `+`
      : invert
      ? `+`
      : `-`;
return `${sign}${hours}:${minutes}`
```
But it's up to you ;) You can ignore my comment, if you like.

File: projects/storefrontlib/src/shared/components/date-picker/date-picker-formatter.service.ts (hunk 9 above)
Comment (Zeyber) [reply projects/storefrontlib/src/shared/components/date-picker/date-picker-formatter.service.ts:21]: Changed when verifying if the date picker is using the correct logic or not as part of the issue related to this pr. Although not part of the final solution, this logic has been corrected here.

Comment (Platonn) [reply projects/storefrontlib/src/shared/components/date-picker/date-picker-formatter.service.ts:21]: It looks like an important bug fix. And it's not related to B2B epic. Then IMO we should create a separate bug-ticket for it and merge it separately to develop (and maintenance branch). It will help leave trace for this bug, trace for the bugfix commit and it will leave a trace in the change log (so it's communicated to customers in which version it was fixed).

Comment (Zeyber) [reply projects/storefrontlib/src/shared/components/date-picker/date-picker-formatter.service.ts:21]: This is actually only added in the b2b epic as the previous way to select start and end dates. There were issues with timezone discrepancies which is why we opted to use the datetime-picker instead.

**Review by Platonn — APPROVED** [review]

File: projects/storefrontlib/src/shared/components/date-picker/date-picker-formatter.service.ts (hunk 9 above)
Comment (Platonn) [reply projects/storefrontlib/src/shared/components/date-picker/date-picker-formatter.service.ts:21]: My bad. 👀 I thought this service was already merged to develop.
